pyop2/codegen/rep2loopy.py

```diff
@@ -0,0 +1,781 @@
+
+    wrapper = loopy.make_kernel(domains,
+                                statements,
+                                kernel_data=parameters.kernel_data,
+                                target=loopy.CTarget(),
+                                temporary_variables=parameters.temporaries,
+                                symbol_manglers=[symbol_mangler],
+                                options=options,
+                                assumptions=assumptions,
+                                lang_version=(2018, 2),
+                                name=wrapper_name)
+
+    # additional assumptions
+    if builder.single_cell:
+        wrapper = loopy.assume(wrapper, "start < end")
+    else:
+        wrapper = loopy.assume(wrapper, "start <= end")
+    wrapper = loopy.assume(wrapper, "start >= 0")
+    if builder.extruded:
+        wrapper = loopy.assume(wrapper, "{0} <= {1}".format(parameters.layer_start, parameters.layer_end))
+
+    # prioritize loops
+    for indices in context.index_ordering:
+        wrapper = loopy.prioritize_loops(wrapper, indices)
+
+    # register kernel
+    kernel = builder.kernel
+    headers = set(kernel._headers)
+    headers = headers | set(["#include <math.h>"])
+    preamble = "\n".join(sorted(headers))
+
+    from coffee.base import Node
+
+    if isinstance(kernel._code, loopy.LoopKernel):
+        knl = kernel._code
+        wrapper = loopy.register_callable_kernel(wrapper, knl)
+        from loopy.transform.callable import _match_caller_callee_argument_dimension_
+        wrapper = _match_caller_callee_argument_dimension_(wrapper, knl.name)
+        wrapper = loopy.inline_callable_kernel(wrapper, knl.name)
+    else:
+        # kernel is a string, add it to preamble
+        if isinstance(kernel._code, Node):
+            code = kernel._code.gencode()
+        else:
+            code = kernel._code
+        wrapper = loopy.register_function_id_to_in_knl_callable_mapper(
+            wrapper,
+            PyOP2KernelLookup(kernel.name, code, tuple(builder.argument_accesses)))
+        preamble = preamble + "\n" + code
+
+    wrapper = loopy.register_preamble_generators(wrapper, [_PreambleGen(preamble)])
+
+    # register petsc functions
+    wrapper = loopy.register_function_id_to_in_knl_callable_mapper(
+        wrapper, petsc_function_lookup)
+
+    return wrapper
+
+
+def argtypes(kernel):
+    args = []
+    for arg in kernel.args:
+        if isinstance(arg, loopy.ValueArg):
+            args.append(as_ctypes(arg.dtype))
+        elif isinstance(arg, loopy.ArrayArg):
+            args.append(ctypes.c_voidp)
+        else:
+            raise ValueError("Unhandled arg type '%s'" % type(arg))
+    return args
+
+
+@singledispatch
+def statement(expr, context):
+    raise AssertionError("Unhandled statement type '%s'" % type(expr))
+
+
+@statement.register(DummyInstruction)
+def statement_dummy(expr, context):
+    new_children = tuple(expression(c, context.parameters) for c in expr.children)
```

Comment:
OK, so this just traverses the children to "register" the loop indices the dummy instruction contains.